File: frontend/src/types/processKnowledge.ts
export type RecordType =
  | "control_narrative"
  | "io_list"
  | "cause_effect_matrix"
  | "alarm_rationalization"
  | "moc"
  | "knowledge_issue"
  | "engineering_note";

export type RecordStatus = "active" | "needing_setup" | "needs_review" | "archived";
export type RevisionStatus = "draft" | "proposed" | "approved" | "rejected" | "archived";
export type ReviewStatus =
  | "open"
  | "approved"
  | "rejected"
  | "needs_manual_review"
  | "resolved";

export interface ProcessUnit {
  id: string;
  name: string;
  area?: string | null;
  description?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ProcessUnitListResponse {
  items: ProcessUnit[];
  total: number;
}

export interface EquipmentModule {
  id: string;
  process_unit_id: string;
  name: string;
  module_type?: string | null;
  description?: string | null;
  aliases: string[];
  metadata_json: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface ControlImportSource {
  id: string;
  process_unit_id: string;
  module_id?: string | null;
  name: string;
  source_system: string;
  acquisition_mode: string;
  connector_hint?: string | null;
  endpoint_url?: string | null;
  export_path?: string | null;
  schedule?: string | null;
  is_enabled: boolean;
  config_json: Record<string, unknown>;
  last_run_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface LogicSnapshot {
  id: string;
  module_id: string;
  source_type: string;
  source_filename: string;
  file_hash?: string | null;
  export_timestamp?: string | null;
  parsed_extract: Record<string, unknown>;
  raw_project_id?: string | null;
  notes?: string | null;
  created_by?: string | null;
  created_at: string;
}

export interface EngineeringRecord {
  id: string;
  process_unit_id: string;
  module_id?: string | null;
  record_type: RecordType;
  title: string;
  status: RecordStatus;
  owner?: string | null;
  description?: string | null;
  metadata_json: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface DocumentRevision {
  id: string;
  engineering_record_id: string;
  source_logic_snapshot_id?: string | null;
  template_id?: string | null;
  revision: string;
  status: RevisionStatus;
  body_markdown?: string | null;
  structured_content: Record<string, unknown>;
  created_by?: string | null;
  approved_by?: string | null;
  approved_at?: string | null;
  created_at: string;
}

export interface LogicDiff {
  id: string;
  module_id: string;
  previous_snapshot_id: string;
  current_snapshot_id: string;
  changed: boolean;
  changed_routines: Record<string, unknown>[];
  changed_tags: Record<string, unknown>[];
  changed_reads_writes: Record<string, unknown>[];
  summary_payload: Record<string, unknown>;
  impacted_record_ids: string[];
  created_at: string;
}

export interface ProposedDocumentUpdate {
  id: string;
  engineering_record_id: string;
  source_logic_snapshot_id: string;
  source_logic_diff_id?: string | null;
  base_document_revision_id?: string | null;
  proposed_body?: string | null;
  proposed_structured_changes: Record<string, unknown>;
  confidence?: string | null;
  status: string;
  created_at: string;
}

export interface ReviewItem {
  id: string;
  process_unit_id: string;
  module_id?: string | null;
  engineering_record_id?: string | null;
  logic_snapshot_id?: string | null;
  logic_diff_id?: string | null;
  proposed_document_update_id?: string | null;
  title: string;
  reason?: string | null;
  status: ReviewStatus;
  assigned_to?: string | null;
  created_at: string;
  resolved_at?: string | null;
}

export interface ApprovalHistory {
  id: string;
  review_item_id?: string | null;
  engineering_record_id?: string | null;
  document_revision_id?: string | null;
  proposed_document_update_id?: string | null;
  decision: "approved" | "rejected" | "commented";
  actor: string;
  comments?: string | null;
  created_at: string;
}

export interface ImportSyncResult {
  run: {
    id: string;
    source_id: string;
    module_id?: string | null;
    status: string;
    acquired_filename?: string | null;
    project_id?: string | null;
    previous_project_id?: string | null;
    snapshot_id?: string | null;
    diff_summary?: string | null;
    impacted_records: Record<string, unknown>[];
    error_message?: string | null;
    started_at: string;
    completed_at?: string | null;
  };
  snapshot?: LogicSnapshot | null;
  diff?: Record<string, unknown> | null;
  logic_diff?: LogicDiff | null;
  impacted_records: Record<string, unknown>[];
  process_unit_id?: string | null;
  module_id?: string | null;
  import_source_id?: string | null;
  snapshot_id?: string | null;
  diff_id?: string | null;
  changed: boolean;
  review_item_ids: string[];
  affected_record_ids: string[];
}

export interface ModuleRecord {
  module: EquipmentModule;
  snapshots: LogicSnapshot[];
  narratives: unknown[];
  governance_artifacts: unknown[];
  issues: unknown[];
}

File: backend/app/schemas/process_knowledge.py
"""API schemas for the process-unit control knowledge system."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.db.models.process_knowledge import (
    ApprovalDecision,
    DocumentStatus,
    DocumentRevisionStatus,
    EngineeringRecordStatus,
    EngineeringRecordType,
    GovernanceArtifactType,
    ImportAcquisitionMode,
    ImportRunStatus,
    IssueStatus,
    ProposedUpdateStatus,
    ReviewItemStatus,
    SnapshotSourceType,
)


class ProcessUnitCreate(BaseModel):
    name: str
    area: str | None = None
    description: str | None = None


class ProcessUnitRead(ProcessUnitCreate):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class ProcessUnitListResponse(BaseModel):
    items: list[ProcessUnitRead]
    total: int


class EquipmentModuleCreate(BaseModel):
    name: str
    module_type: str | None = None
    description: str | None = None
    aliases: list[str] = Field(default_factory=list)
    metadata_json: dict[str, Any] = Field(default_factory=dict)


class EquipmentModuleRead(EquipmentModuleCreate):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    process_unit_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class LogicSnapshotCreate(BaseModel):
    source_type: SnapshotSourceType = SnapshotSourceType.XML
    source_filename: str
    file_hash: str | None = None
    export_timestamp: datetime | None = None
    parsed_extract: dict[str, Any] = Field(default_factory=dict)
    raw_project_id: str | None = None
    notes: str | None = None
    created_by: str | None = None


class LogicSnapshotRead(LogicSnapshotCreate):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    module_id: uuid.UUID
    created_at: datetime


class ControlNarrativeCreate(BaseModel):
    snapshot_id: uuid.UUID | None = None
    title: str
    body_markdown: str
    status: DocumentStatus = DocumentStatus.DRAFT
    revision: str = "0.1"
    owner: str | None = None
    reviewer: str | None = None
    approved_by: str | None = None
    approved_at: datetime | None = None
    metadata_json: dict[str, Any] = Field(default_factory=dict)


class ControlNarrativeRead(ControlNarrativeCreate):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    module_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class GovernanceArtifactCreate(BaseModel):
    snapshot_id: uuid.UUID | None = None
    artifact_type: GovernanceArtifactType
    title: str
    status: DocumentStatus = DocumentStatus.DRAFT
    content_json: dict[str, Any] = Field(default_factory=dict)
    source_filename: str | None = None
    owner: str | None = None
    reviewer: str | None = None
    approved_by: str | None = None
    approved_at: datetime | None = None


class GovernanceArtifactRead(GovernanceArtifactCreate):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    module_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class ControlIssueCreate(BaseModel):
    found_in_snapshot_id: uuid.UUID | None = None
    fixed_in_snapshot_id: uuid.UUID | None = None
    title: str
    symptom: str | None = None
    root_cause: str | None = None
    fix_description: str | None = None
    severity: str | None = None
    status: IssueStatus = IssueStatus.OPEN
    tags: list[str] = Field(default_factory=list)
    evidence: list[dict[str, Any]] = Field(default_factory=list)
    created_by: str | None = None
    verified_by: str | None = None


class ControlIssueRead(ControlIssueCreate):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    module_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class ControlImportSourceCreate(BaseModel):
    process_unit_id: uuid.UUID
    module_id: uuid.UUID | None = None
    name: str
    source_system: str
    acquisition_mode: ImportAcquisitionMode = ImportAcquisitionMode.MANUAL_UPLOAD
    connector_hint: str | None = None
    endpoint_url: str | None = None
    export_path: str | None = None
    schedule: str | None = None
    is_enabled: bool = True
    config_json: dict[str, Any] = Field(default_factory=dict)


class ControlImportSourceRead(ControlImportSourceCreate):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    last_run_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class ControlImportRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    source_id: uuid.UUID
    module_id: uuid.UUID | None = None
    status: ImportRunStatus
    acquired_filename: str | None = None
    project_id: str | None = None
    previous_project_id: str | None = None
    snapshot_id: uuid.UUID | None = None
    diff_summary: str | None = None
    impacted_records: list[dict[str, Any]] = Field(default_factory=list)
    error_message: str | None = None
    started_at: datetime
    completed_at: datetime | None = None


class EngineeringRecordCreate(BaseModel):
    process_unit_id: uuid.UUID
    module_id: uuid.UUID | None = None
    record_type: EngineeringRecordType
    title: str
    status: EngineeringRecordStatus = EngineeringRecordStatus.ACTIVE
    owner: str | None = None
    description: str | None = None
    metadata_json: dict[str, Any] = Field(default_factory=dict)


class EngineeringRecordRead(EngineeringRecordCreate):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class DocumentTemplateCreate(BaseModel):
    name: str
    record_type: EngineeringRecordType
    scope_company: str | None = None
    scope_site: str | None = None
    template_body: str | None = None
    template_schema: dict[str, Any] = Field(default_factory=dict)
    active: bool = True
    version: str = "1.0"


class DocumentTemplateRead(DocumentTemplateCreate):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class DocumentRevisionCreate(BaseModel):
    source_logic_snapshot_id: uuid.UUID | None = None
    template_id: uuid.UUID | None = None
    revision: str
    status: DocumentRevisionStatus = DocumentRevisionStatus.DRAFT
    body_markdown: str | None = None
    structured_content: dict[str, Any] = Field(default_factory=dict)
    created_by: str | None = None
    approved_by: str | None = None
    approved_at: datetime | None = None


class DocumentRevisionRead(DocumentRevisionCreate):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    engineering_record_id: uuid.UUID
    created_at: datetime


class LogicDiffRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    module_id: uuid.UUID
    previous_snapshot_id: uuid.UUID
    current_snapshot_id: uuid.UUID
    changed: bool
    changed_routines: list[dict[str, Any]] = Field(default_factory=list)
    changed_tags: list[dict[str, Any]] = Field(default_factory=list)
    changed_reads_writes: list[dict[str, Any]] = Field(default_factory=list)
    summary_payload: dict[str, Any] = Field(default_factory=dict)
    impacted_record_ids: list[str] = Field(default_factory=list)
    created_at: datetime


class ProposedDocumentUpdateRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    engineering_record_id: uuid.UUID
    source_logic_snapshot_id: uuid.UUID
    source_logic_diff_id: uuid.UUID | None = None
    base_document_revision_id: uuid.UUID | None = None
    proposed_body: str | None = None
    proposed_structured_changes: dict[str, Any] = Field(default_factory=dict)
    confidence: str | None = None
    status: ProposedUpdateStatus
    created_at: datetime


class ReviewItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    process_unit_id: uuid.UUID
    module_id: uuid.UUID | None = None
    engineering_record_id: uuid.UUID | None = None
    logic_snapshot_id: uuid.UUID | None = None
    logic_diff_id: uuid.UUID | None = None
    proposed_document_update_id: uuid.UUID | None = None
    title: str
    reason: str | None = None
    status: ReviewItemStatus
    assigned_to: str | None = None
    created_at: datetime
    resolved_at: datetime | None = None


class ReviewDecisionRequest(BaseModel):
    actor: str
    comments: str | None = None


class ApprovalHistoryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    review_item_id: uuid.UUID | None = None
    engineering_record_id: uuid.UUID | None = None
    document_revision_id: uuid.UUID | None = None
    proposed_document_update_id: uuid.UUID | None = None
    decision: ApprovalDecision
    actor: str
    comments: str | None = None
    created_at: datetime


class ImportSyncResult(BaseModel):
    run: ControlImportRunRead
    snapshot: LogicSnapshotRead | None = None
    diff: dict[str, Any] | None = None
    logic_diff: LogicDiffRead | None = None
    impacted_records: list[dict[str, Any]] = Field(default_factory=list)
    process_unit_id: uuid.UUID | None = None
    module_id: uuid.UUID | None = None
    import_source_id: uuid.UUID | None = None
    snapshot_id: uuid.UUID | None = None
    diff_id: uuid.UUID | None = None
    changed: bool = False
    review_item_ids: list[str] = Field(default_factory=list)
    affected_record_ids: list[str] = Field(default_factory=list)


class ModuleRecordResponse(BaseModel):
    module: EquipmentModuleRead
    snapshots: list[LogicSnapshotRead]
    narratives: list[ControlNarrativeRead]
    governance_artifacts: list[GovernanceArtifactRead]
    issues: list[ControlIssueRead]
